File: api/crs/codegen/lib/crs_prodAtx/crs_prodAtx.c
#include "crs_prodAtx.h"
#include "m2c.h"
#include "omp.h"
#include <math.h>

static void b_crs_prodAtx(const emxArray_int32_T *A_row_ptr, const
  emxArray_int32_T *A_col_ind, const emxArray_real_T *A_val, int A_nrows, int
  A_ncols, const emxArray_real_T *x, emxArray_real_T *b);
static void b_crs_prodAtx_kernel(const emxArray_int32_T *row_ptr, const
  emxArray_int32_T *col_ind, const emxArray_real_T *val, const emxArray_real_T
  *x, int x_m, emxArray_real_T *b, int b_m, int nrows, int ncols, int nrhs);
static void c_crs_prodAtx(const emxArray_int32_T *A_row_ptr, const
  emxArray_int32_T *A_col_ind, const emxArray_real_T *A_val, int A_nrows, int
  A_ncols, const emxArray_real_T *x, emxArray_real_T *b);
static void crs_prodAtx_kernel(const emxArray_int32_T *row_ptr, const
  emxArray_int32_T *col_ind, const emxArray_real_T *val, const emxArray_real_T
  *x, int x_m, emxArray_real_T *b, int b_m, int nrows, int ncols, int nrhs,
  boolean_T varargin_1);
static void m2c_error(void);
static void m2c_warn(void);
static void b_crs_prodAtx(const emxArray_int32_T *A_row_ptr, const
  emxArray_int32_T *A_col_ind, const emxArray_real_T *A_val, int A_nrows, int
  A_ncols, const emxArray_real_T *x, emxArray_real_T *b)
{
  int i;
  i = b->size[0] * b->size[1];
  b->size[0] = A_ncols;
  b->size[1] = x->size[1];
  emxEnsureCapacity_real_T(b, i);
  b_crs_prodAtx_kernel(A_row_ptr, A_col_ind, A_val, x, x->size[0], b, A_ncols,
                       A_nrows, A_ncols, x->size[1]);
}

static void b_crs_prodAtx_kernel(const emxArray_int32_T *row_ptr, const
  emxArray_int32_T *col_ind, const emxArray_real_T *val, const emxArray_real_T
  *x, int x_m, emxArray_real_T *b, int b_m, int nrows, int ncols, int nrhs)
{
  int boffset;
  int xoffset;
  int k;
  int i;
  int i1;
  int j;
  int b_i;
  double alpha;
  int r;
  boffset = 0;
  if (1 <= nrows) {
    xoffset = 0;
    for (k = 0; k < nrhs; k++) {
      i = boffset + 1;
      i1 = boffset + ncols;
      for (j = i; j <= i1; j++) {
        b->data[j - 1] = 0.0;
      }

      for (b_i = 0; b_i < nrows; b_i++) {
        alpha = x->data[b_i + xoffset];
        i = row_ptr->data[b_i];
        i1 = row_ptr->data[b_i + 1] - 1;
        for (j = i; j <= i1; j++) {
          r = (boffset + col_ind->data[j - 1]) - 1;
          b->data[r] += alpha * val->data[j - 1];
        }
      }

      xoffset += x_m;
      boffset += b_m;
    }
  }
}

static void c_crs_prodAtx(const emxArray_int32_T *A_row_ptr, const
  emxArray_int32_T *A_col_ind, const emxArray_real_T *A_val, int A_nrows, int
  A_ncols, const emxArray_real_T *x, emxArray_real_T *b)
{
  if ((b->size[0] < A_ncols) || (b->size[1] < x->size[1])) {
    m2c_error();
  }

  b_crs_prodAtx_kernel(A_row_ptr, A_col_ind, A_val, x, x->size[0], b, b->size[0],
                       A_nrows, A_ncols, x->size[1]);
}

static void crs_prodAtx_kernel(const emxArray_int32_T *row_ptr, const
  emxArray_int32_T *col_ind, const emxArray_real_T *val, const emxArray_real_T
  *x, int x_m, emxArray_real_T *b, int b_m, int nrows, int ncols, int nrhs,
  boolean_T varargin_1)
{
  int nthreads;
  int boffset;
  int xoffset;
  int istart;
  int iend;
  int k;
  int chunk;
  int i;
  int j;
  int b_i;
  int b_remainder;
  double alpha;
  if (varargin_1) {
    xoffset = omp_get_num_threads();
    nthreads = (int)floor((double)b_m / (double)ncols);
    if (xoffset < nthreads) {
      nthreads = xoffset;
    }

    xoffset = omp_get_thread_num();
    boffset = xoffset * ncols;
    if (nthreads == 1) {
      istart = 0;
      iend = nrows;
    } else {
      iend = omp_get_thread_num();
      if (iend >= nthreads) {
        istart = 0;
        iend = 0;
      } else {
        chunk = nrows / nthreads;
        b_remainder = nrows - nthreads * chunk;
        if (b_remainder < iend) {
          xoffset = b_remainder;
        } else {
          xoffset = iend;
        }

        istart = iend * chunk + xoffset;
        iend = (istart + chunk) + (iend < b_remainder);
      }
    }

    istart++;
  } else {
    nthreads = 1;
    boffset = 0;
    istart = 1;
    iend = nrows;
  }

  if (istart <= iend) {
    xoffset = -1;
    for (k = 0; k < nrhs; k++) {
      chunk = boffset + 1;
      i = boffset + ncols;
      for (j = chunk; j <= i; j++) {
        b->data[j - 1] = 0.0;
      }

      for (b_i = istart; b_i <= iend; b_i++) {
        alpha = x->data[b_i + xoffset];
        chunk = row_ptr->data[b_i - 1];
        i = row_ptr->data[b_i] - 1;
        for (j = chunk; j <= i; j++) {
          b_remainder = (boffset + col_ind->data[j - 1]) - 1;
          b->data[b_remainder] += alpha * val->data[j - 1];
        }
      }

      xoffset += x_m;
      boffset += b_m;
    }
  }

  if (nthreads > 1) {

#pragma omp barrier

    xoffset = omp_get_num_threads();
    if (xoffset == 1) {
      istart = 0;
      iend = ncols;
    } else {
      iend = omp_get_thread_num();
      chunk = ncols / xoffset;
      b_remainder = ncols - xoffset * chunk;
      if (b_remainder < iend) {
        xoffset = b_remainder;
      } else {
        xoffset = iend;
      }

      istart = iend * chunk + xoffset;
      iend = (istart + chunk) + (iend < b_remainder);
    }

    xoffset = ncols;
    for (j = 2; j <= nthreads; j++) {
      boffset = 1;
      for (k = 0; k < nrhs; k++) {
        chunk = boffset + istart;
        i = (boffset + iend) - 1;
        for (b_i = chunk; b_i <= i; b_i++) {
          b->data[b_i - 1] += b->data[(xoffset + b_i) - 1];
        }

        boffset += b_m;
      }

      xoffset += ncols;
    }
  }
}

static void m2c_error(void)
{
  M2C_error("crs_prodAtx:BufferTooSmal",
            "Buffer space for output b is too small.");
}

static void m2c_warn(void)
{
  M2C_warn("crs_prodAtx:NestedParallel",
           "You are trying to use nested parallel regions. Solution may be incorrect.");
}

void crs_prodAtx(const struct0_T *A, const emxArray_real_T *x, emxArray_real_T
                 *b, const emxArray_int32_T *nthreads)
{
  int n;
  if ((b->size[0] < A->ncols) || (b->size[1] < x->size[1])) {
    m2c_error();
  }

  if (nthreads->size[0] != 0) {
    n = omp_get_nested();
    if (n == 0) {
      n = omp_get_num_threads();
      if ((n > 1) && (nthreads->data[0] > 1)) {

#pragma omp master
        {
          m2c_warn();
        }

      }
    }

#pragma omp parallel default(shared) num_threads(nthreads->data[0])
    {
      n = omp_get_num_threads();
      crs_prodAtx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], b, b->
                         size[0], A->nrows, A->ncols, x->size[1], n > 1);
    }

  } else {
    n = omp_get_num_threads();
    if (n == 1) {
      b_crs_prodAtx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], b,
                           b->size[0], A->nrows, A->ncols, x->size[1]);
    } else {
      crs_prodAtx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], b,
                         b->size[0], A->nrows, A->ncols, x->size[1], true);
    }
  }
}

void crs_prodAtx_initialize(void)
{
}

void crs_prodAtx_ser(const struct0_T *A, const emxArray_real_T *x,
                     emxArray_real_T *b)
{
  b_crs_prodAtx(A->row_ptr, A->col_ind, A->val, A->nrows, A->ncols, x, b);
}

void crs_prodAtx_ser1(const struct0_T *A, const emxArray_real_T *x,
                      emxArray_real_T *b)
{
  c_crs_prodAtx(A->row_ptr, A->col_ind, A->val, A->nrows, A->ncols, x, b);
}

void crs_prodAtx_terminate(void)
{
}

File: api/crs/crs_prodAtx.m
function b = crs_prodAtx(A, x, b, nthreads, varargin)
%crs_prodAtx Compute b=A'*x for a sparse matrix A in CRS format,
% assuming that number of columns of A is equal to size(x,1).
%
%      b = crs_prodAtx(A, x [, b])
% Computes b=A'*x in serial.
%
%      b = crs_prodAtx(A, x, b, nthreads)
% Computes b=A'*x using nthreads OpenMP threads, where nthreads is an int32.
% In this mode, there are implicit barriers in the function.
% It will use min(nthreads, size(b,1)/A.ncols, ompGetMaxThreads) threads.
%
%      b = crs_prodAtx(A, x, b, [])
% Computes b=A'*x locally within each OpenMP thread, assuming the parallel
% region has already been initialized. The argument b has been preallocated.
% It will use min(size(b,1)/A.ncols, ompGetNumThreads)
% threads forcomputations. In this mode, there is no implicit barrier.
%
%      b = crs_prodAtx(A, x, b, nthreads, comm)
% also performs communication within the MPI communicator.
% It assumes that A is partitioned rowwise, and hence A'*x is
% a partial sum on each process. An allreduce is performed on b
% within the communicator.
%
% If A is partitioned columnwise, then b must be duplicated on all 
% processes, and crs_prodAtx should be called without the communicator.
%
%      b = crs_prodAtx(A, x, b, nthreads, comm, pgmsg [, pgsz])
% specifies an additiona message that should be summed during
% allreduce and strored in b. This is to piggy-back on allreduce
% to help aggregate small messages to optimize MPI communication.
%
% The compiled version reqiures A and x to be double precision.
% The M file works for both single and double precision.

% See http://www.netlib.org/linalg/html_templates/node98.html

%#codegen -args {crs_matrix, coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]), 
%#codegen coder.typeof(int32(1), [1,1], [1,0])}
%#codegen crs_prodAtx_ser -args {crs_matrix, coder.typeof(0, [inf,inf])}
%#codegen crs_prodAtx_ser1 -args {crs_matrix, coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf])}
% %#codegen crs_prodAtx_mpi -args {crs_matrix, coder.typeof(0, [inf,inf]),
% %#codegen coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0]), opaque_obj}
% %#codegen crs_prodAtx_mpip -args {crs_matrix, coder.typeof(0, [inf,inf]),
% %#codegen coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0]), opaque_obj, coder.typeof(0, [inf,1])}
% %#codegen crs_prodAtx_mpip1 -args {crs_matrix, coder.typeof(0, [inf,inf]),
% %#codegen coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0]), opaque_obj,coder.typeof(0, [inf,1]), int32(0)}

coder.inline('never');

if nargin==2;
    b = coder.nullcopy(zeros(A.ncols,size(x,2)));
else
    if size(b,1)<A.ncols || size(b,2)<size(x,2)
        m2c_error('crs_prodAtx:BufferTooSmal', 'Buffer space for output b is too small.');
    end
end
ismt = nargin>3 && ompGetNumThreads>1;

if nargin>3 && ~isempty(nthreads)
    %% Declare parallel region
    if ~ompGetNested && ismt && nthreads(1)>1
        OMP_begin_master
        m2c_warn('crs_prodAtx:NestedParallel', ...
            'You are trying to use nested parallel regions. Solution may be incorrect.');
        OMP_end_master
    end
    
    OMP_begin_parallel(nthreads(1));

    %% Compute b=A'*x
    b = crs_prodAtx_kernel(A.row_ptr, A.col_ind, A.val, x, int32(size(x,1)), ...
        b, int32(size(b,1)), A.nrows, A.ncols, int32(size(x,2)), ompGetNumThreads>1);
    
    OMP_end_parallel
else
    %% Compute b=A'*x
    b = crs_prodAtx_kernel(A.row_ptr, A.col_ind, A.val, x, int32(size(x,1)), ...
        b, int32(size(b,1)), A.nrows, A.ncols, int32(size(x,2)), ismt);

    if ~isempty(varargin) && ismt; OMP_barrier; end
end

if ~isempty(varargin)
    % Perform MPI allreduce
    OMP_begin_single
    s = int32(A.ncols*size(x,2));
    b = allreduce(b, s, MPI_SUM, varargin{:});
    OMP_end_single
end

function b = crs_prodAtx_kernel(row_ptr, col_ind, val, x, x_m, ...
    b, b_m, nrows, ncols, nrhs, ismt)

coder.inline('never');

if ismt
    nthreads = min(ompGetNumThreads, ...
        int32(floor(double(b_m)/double(ncols))));
    boffset = ompGetThreadNum*ncols;
    [istart, iend] = OMP_local_chunk(nrows, nthreads);
else
    nthreads = int32(1); boffset = int32(0);
    istart = int32(1); iend = int32(nrows);
end

if istart<=iend
    xoffset = int32(0);
    for k=1:nrhs
        for j=boffset+1:boffset+ncols; b(j) = 0; end
        
        for i=istart:iend
            alpha = x(i+xoffset);
            
            for j = row_ptr(i) : row_ptr(i+1) - 1
                r = boffset+col_ind(j);
                b(r) = b(r) + alpha * val(j);
            end
        end
        
        xoffset = xoffset + x_m; boffset = boffset + b_m;
    end
end

%% Perfom summation of partial sums in b
if nthreads>1
    OMP_barrier;
    [istart, iend] = OMP_local_chunk(ncols);
    
    offset = ncols;
    for j=2:nthreads
        boffset = int32(0);
        for k=1:nrhs
            for i=boffset+istart:boffset+iend
                b(i) = b(i) + b(offset+i);
            end
            boffset = boffset + b_m;
        end
        offset = offset + ncols;
    end
end

function test %#ok<DEFNU>
%!test
%! if ~exist(['crs_prodAtx.' mexext], 'file')
%!    m=100; n = 20;
%! else
%!    m=10000; n = 2000;
%! end
%! tic; sp = sprand(m,n,0.5); x=rand(size(sp,1),2);
%! [is,js,vs] = find(sp);
%! fprintf(1, '\n\tGenerated random matrix in %g seconds\n', toc);
%! tic; b0 = sp'*x;
%! fprintf(1, '\tComputed reference solution in %g seconds\n', toc);
%! tic; A = crs_matrix(int32(is), int32(js), vs, int32(size(sp,1)), int32(size(sp,2)));
%! fprintf(1, '\tConverted into crs_matrix in %g seconds\n', toc);
%! fprintf(1, '\tTesting serial: ');
%! tic; b1 = crs_prodAtx(A, x);
%! fprintf(1, 'Done in %g seconds\n ', toc);
%! assert(norm(b0-b1)/norm(b0)<=1.e-10);

%! for nthreads=int32([1 2 4 8])
%!     if nthreads>ompGetMaxThreads; break; end
%!     fprintf(1, '\tTesting %d thread(s): ', nthreads);
%!     tic; b2 = zeros(size(sp,2)*nthreads,2);
%!     b2 = crs_prodAtx(A, x, b2, nthreads);
%!     fprintf(1, 'Done in %g seconds\n ', toc);
%!     assert(norm(b0-b2(1:size(sp,2),:))/norm(b0)<=1.e-10);
%! end

% %! nprocs = double(comm_size(MPI_COMM_WORLD));
% %! fprintf(1, '\tTesting 2 threads with MPI call: ');
% %! b2 = zeros(size(sp,2)*2,2);
% %! tic; b2 = crs_prodAtx(A, x, b2, int32(2), MPI_COMM_WORLD);
% %! fprintf(1, 'Done in %g seconds\n ', toc);
% %! assert(norm(b0-b2(1:size(sp,2),:))/norm(b0)<=1.e-10);
% 
% %! b3 = zeros(2*size(sp,2)+1,1);
% %! fprintf(1, '\tTesting 2 threads with piggy-back: ');
% %! tic; b3 = crs_prodAtx(A, x(:,1), b3, int32(2), MPI_COMM_WORLD, 1, int32(1));
% %! fprintf(1, 'Done in %g seconds\n ', toc);
% %! assert(nprocs>1 || norm(b0(:,1)-b3(1:size(sp,2)))/norm(b0)<=1.e-10);
% %! assert(b3(size(sp,2)+1)==nprocs);
